### ISSUE TITLE
twisted.web.http_headers.Headers allows itself to be constructed from strings instead of lists

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4022]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4022|
|Type|enhancement|
|Created|2009-09-14 15:45:05Z|

`Headers` is basically an abstraction around a dict mapping `str` to `list` of `str`.  So its initializer will accept such a dict.  However, it will also happily accept a dict mapping `str` to `str` and populate itself with garbage.  It should reject this case.

Attachments:

* [strict-headers-constructor-4022.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e1a/e1a89047db4bc61c961ad0424dd44a19b6d4cbe3/9ff5f58bb3653cc7765033b759f9fa0b590eb284.patch) (2253 bytes) - added by rwall on 2009-11-07 17:29:24Z - Add validation for the addHeader value argument and refactor the setHeaders method to use it.
* [strict-setRawHeaders-4022.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e1a/e1a89047db4bc61c961ad0424dd44a19b6d4cbe3/f52d1ac4d2ef3e6fce73c9291095d44f72815c1b.diff) (1136 bytes) - added by djfroofy on 2010-02-22 18:47:58Z - Patch to only accept list value in setRawHeaders

<details><summary>Searchable metadata</summary>

```
trac-id__4022 4022
type__enhancement enhancement
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1252943105000000 1252943105000000
changetime__1266961488000000 1266961488000000
version__None None
owner__ 
cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj set owner to @exarkun|
|-|-|

This looked easy enough so I had a go. Patch attached.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @wallrj|
|-|-|

Thanks.  It looks like the patch is doing a bunch of things beyond just what's described on this ticket:

  1. What's the purpose for the change to `__init__`?
  1. Same question for `addRawHeader`
  1. The change that looks like it actually addresses this ticket, in `setRawHeaders`, looks like it's actually doing three things:
    1. It rejects `str` values (ostensibly the change desired by this ticket)
    1. It rejects `unicode` values too.  This is cool, but there's no unit test.
    1. It converts any non-str, non-unicode iterable into a list, whereas the previous code didn't.  This might be a good change... I'm not sure.  This is the low-level API.  The primary purpose isn't to be friendly.  Accepting a whole mess of different types might not be called for.  At the very least, it also needs documentation and testing.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_3"></a>|@djfroofy set owner to @exarkun|
|-|-|

I'm personally not a fan of explicitly type checking if the documentation around accepted types is clear.  Anyhow, patch attached.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `assigned`|
|-|-|

Thanks, looks good to me, I'll apply.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [28479]) Apply strict-setRawHeaders-4022.diff adding type checking to Header's initializer 

Author: djfroofy
Reviewer: exarkun
Fixes: [#4022](https://github.com/twisted/twisted/issues/4022)

Reject any values passed to twisted.web.http_headers.Headers.setRawHeaders which are not lists.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
